ballerina/jwt_oauth2_handler.bal

```diff
@@ -0,0 +1,96 @@
+//
+// WSO2 LLC. licenses this file to you under the Apache License,
+// Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/http;
+import ballerina/jwt;
+
+const string OAUTH_URL = "https://oauth2.googleapis.com";
+
+# Defines the OAuth2 handler for client authentication.
+isolated class ClientOAuth2ExtensionGrantHandler {
+    private final OAuth2ExtensionGrantProvider provider;
+
+    public isolated function init(jwt:IssuerConfig? config = ()) returns error? {
+        self.provider = check new (config);
+        return;
+    }
+
+    public isolated function getSecurityHeaders(string? userAccount, map<string|string[]> headers) returns error? {
+        if userAccount !is string {
+            return error("User account is not provided.");
+        }
+        string|error result = self.provider.generateToken(userAccount);
+        if result is string {
+            headers[http:AUTH_HEADER] = http:AUTH_SCHEME_BEARER + " " + result;
+        } else {
+            return error("Failed to enrich headers with OAuth2 token.", result);
+        }
+    }
+
+    isolated function isServiceAccount() returns boolean {
+        return self.provider.getServiceAccountState();
+    }
+}
+
+isolated class OAuth2ExtensionGrantProvider {
+    private jwt:IssuerConfig jwtIssuerConfig = {};
+    private boolean isServiceAccount = false;
+    private final http:Client clientEndpoint;
+
+    public isolated function init(jwt:IssuerConfig? config) returns error? {
+        self.clientEndpoint = check new (OAUTH_URL);
+        if config is jwt:IssuerConfig {
+            self.jwtIssuerConfig = config.clone();
+            self.isServiceAccount = true;
+        }
+        return;
+    }
+
+    public isolated function generateToken(string userAccount) returns string|error {
+        lock {
+            self.jwtIssuerConfig.customClaims["sub"] = userAccount;
+            string|jwt:Error jwt = jwt:issue(self.jwtIssuerConfig);
+            if jwt is string {
+                json payload = {"grant_type": "urn:ietf:params:oauth:grant-type:jwt-bearer", "assertion": jwt};
+                http:Response|http:ClientError response = self.clientEndpoint->post("/token", payload);
+                if response is http:Response {
+                    return extractAccessToken(response);
+                } else {
+                    return error("Failed to call the token endpoint.", response);
+                }
+            }
+            return error("Failed to generate JWT.", <error>jwt);
```

Comment:
why do we need casting here? Can we use an else if clause to check if its an `error`, to avoid casting?